Add product block collection variant for the block email editor

lysyjan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->
- This PR adds a new collection type for the product collection block. The new collection contains a cart content for the user passed in the context.
- There are also some fixes for the email rendering that caused incorrect behavior in the MailPoet plugin.

Closes STOMAIL-7500

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| After |
| ----- |
|<img width="847" height="400" alt="Screenshot 2025-10-24 at 11 29 51" src="https://github.com/user-attachments/assets/e190563d-30dc-4371-bc14-9d2459218081" />|


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Activate the feature Block Email Editor in WooCommerce settings
2. Finish an order for a customer
3. Add the new collection `Cart contents`, for example, in the notification `Order details` that can be manually sent
4. Add some content to the cart for the same user as was used in step 2
5. Sent manually the email `Order details` from WooCommerce Orders for the same user as was used in step 2
6. You should see the cart content in the recieved email

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

lysyjan committed:
Simplify block rendering by directly using `block_content` instead of processing `innerBlocks`

Using the `render_block` function in wrapper blocks caused unintentional rerendering, which affected performance and caused a bug with empty content when the product collection block was used in MailPoet.

lysyjan committed:
Fix used `$post` instead of `$email_post` in backup assignment for proper context

lysyjan committed:
Introduce `get_inner_content` utility to extract inner content

The block_content variable contains a wrapping div and the new method removes it when needed.

lysyjan committed:
Add email-specific context support to rendering context

Introduced an `email_context` property in `Rendering_Context` to provide email-specific data (e.g., user ID, recipient email, order ID).
Updated block rendering to utilize the new context and added tests to validate its behavior.
Added filter `woocommerce_email_editor_rendering_email_context` for extensibility. Updated documentation accordingly.

lysyjan committed:
Add "Cart Contents" collection for email-specific product display

Introduced new "Cart Contents" product collection for use in abandoned cart emails.
Updated backend logic, email rendering, and registration to handle cart-based queries.
Implemented email-specific collection registration and custom query handlers. Added fallback to sample products for preview.

lysyjan committed:
Add email context support to BlockEmailRenderer

lysyjan committed:
Add changelogs

lysyjan committed:
Update cart collection icon to custom cart SVG

lysyjan committed:
Fix empty core/post-content block rendering in batch email processing

Replaced WordPress's static render callback with a stateless implementation to address issues in MailPoet batch email processing. Added integration tests to validate multiple renders per request. Updated block registration and initializer logic to override the default callback.

lysyjan commented on packages/php/email-editor/src/Integrations/Core/Renderer/Blocks/class-quote.php:
```diff
@@ -40,15 +39,9 @@ protected function render_content( string $block_content, array $parsed_block, R
-		}
-
 		return str_replace(
 			array( '{quote_content}', '{citation_content}' ),
-			array( $content, $citation_content ),
+			array( $this->get_inner_content( $block_content ), $citation_content ),
 			$this->get_block_wrapper( $block_content, $parsed_block, $rendering_context )
 		);
```
Thanks for your comment. It works as expected.

costasovo commented on packages/php/email-editor/src/Engine/Renderer/ContentRenderer/class-rendering-context.php:
```diff
@@ -106,4 +122,51 @@ public function translate_slug_to_color( string $color_slug ): string {
+	 */
+	public function get_order_id(): ?int {
+		return isset( $this->email_context['order_id'] ) && is_numeric( $this->email_context['order_id'] ) ? (int) $this->email_context['order_id'] : null;
+	}
```
I don't think that `get_order_id` should be part of the email editor's context.  get_user_id or get_recipient_email are fine as recipient_email is common for all emails and user_id is related to WordPress core, but order is Woo specific so this should be IMO somehow added to the context by Woo. Similarly, woo_email_type is just for Woo notifications. maybe we could live with the generic `get`for now.

costasovo commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/collections/cart-contents.tsx:
```diff
@@ -0,0 +1,78 @@
+	title: __( 'Cart Contents', 'woocommerce' ),
+	icon: <Icon icon={ box } />,
+	description: __(
+		'Display products from the customer cart for abandoned cart emails.',
```
Is the block only for emails, or can it also be used on a page? I mean, it could make sense to add it to an account page...

costasovo replied:
Oh, I see we have email-specific collection, so it is only for emails. Would it be complex to allow it everywhere? It could be a simplification on JS side...

lysyjan replied:
> Oh, I see we have email-specific collection, so it is only for emails. Would it be complex to allow it everywhere? It could be a simplification on JS side...

I think we need to check the design. For example, the item count could be missing. I have checked the existing blocks, and at this moment, there is no similar block. If you think it is worth the effort to have this block available everywhere, I would address it in a new ticket.

costasovo replied:
If it is complex, then a new ticket makes sense. I was thinking it could be a simplification, since we could avoid adding an email-only collection.

costasovo commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/HandlerRegistry.php:
```diff
@@ -378,4 +398,45 @@ function ( $item ) {
+				'product',
+				3
+			)
+		);
```
I think it would be better to use a higher-level method to fetch the IDs. This will protect us in case there are some changes in the DB.
I think this should work:
```php
// IDs only
$ids = wc_get_products( [
    'status'   => 'publish',
    'orderby'  => 'date',
    'order'    => 'DESC',
    'limit'    => 3,
    'return'   => 'ids',   // or 'objects' for WC_Product objects
] );
```

costasovo commented on plugins/woocommerce/src/Internal/EmailEditor/BlockEmailRenderer.php:
```diff
@@ -112,4 +126,34 @@ private function render_block_email( \WP_Post $email_post, string $woo_content,
+	 */
+	private function build_email_context( \WC_Email $wc_email ): array {
+		$recipient_raw = $wc_email->get_recipient();
+		$emails        = array_values( array_filter( array_map( 'sanitize_email', array_map( 'trim', explode( ',', $recipient_raw ) ) ) ) );
+		$context       = array(
+			'recipient_email' => $emails[0] ?? null,
+			'woo_email_type'  => $wc_email->id,
+		);
+
+		// Extract order-related context if the email object is an order.
+		if ( isset( $wc_email->object ) && $wc_email->object instanceof \WC_Order ) {
+			$order               = $wc_email->object;
+			$context['order_id'] = $order->get_id();
+			$context['user_id']  = $order->get_customer_id();
+		}
+
+		return $context;
+	}
```
I remember seeing a similar piece of code and I found it [here](https://github.com/woocommerce/woocommerce/blob/8b1f23220364e796d172c79dbf0ed98a69369520/plugins/woocommerce/src/Internal/EmailEditor/TransactionalEmailPersonalizer.php#L67-L81).
It looks like now we have two separate contexts (for rendering and for personalization tags). Do you think we could unify that and use one context for both?

lysyjan replied:
Yes, we could do it, and it makes sense. Nevertheless, I would prefer a new ticket for this. Do you agree?

costasovo replied:
> I would prefer a new ticket for this. Do you agree?

Sure. It can be handled in a new ticket.

costasovo reviewed (commented):
@lysyjan Thank you for thorough debugging and fixing the renderer issues! I added a couple of comments with some questions and suggestions.

lysyjan committed:
Replace custom cart SVG with WooCommerce icons library asset

STOMAIL-7500

lysyjan committed:
Refactor Rendering_Context to centralize email context access

Replaced specific `get_order_id` method with the generic `get()` for retrieving email context values. Updated related tests and documentation to reflect changes. Enhanced flexibility for extensions by allowing custom context data access.

lysyjan committed:
Remove unused `woo_email_type` and `order_id` context from BlockEmailRenderer

STOMAIL-7500

lysyjan committed:
Refactor to replace custom DB query with `wc_get_products()` for recent products

costasovo reviewed (commented):
@lysyjan Thank you for the updates!

I can confirm that the collections renders in editor, preview and also when order details are sent manually.

I have a question, though. I followed the testing instructions and added the cart collection to the Order details email and I ordered one product. Then I sent the order detail email to the customer. I was expecting that the cart collection would be empty because the customer didn't have anything in the cart. But the email I received contained three items (same as the demo content). So I think that there might be an issue in product fetching for the cart block.

I also noticed a rendering issue. This isn't related only to the cart collection, but I think it happens for the product collections in general. It looks good in the editor, but in preview and in a sent email there is a huge gap below each product image.

<img width="729" height="958" alt="Screenshot 2025-10-29 at 10 43 33" src="https://github.com/user-attachments/assets/293df125-02ee-4bdb-b13c-ad99ad02b32a" />

<img width="719" height="888" alt="Screenshot 2025-10-29 at 10 43 50" src="https://github.com/user-attachments/assets/fb6b9fef-c494-4e34-a271-aef5f58dd68c" />

lysyjan committed:
Fix cart collection empty state handling for emails and previews

lysyjan committed:
Sanitize user input for `userId` and `userEmail`

lysyjan commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/HandlerRegistry.php:
```diff
@@ -378,4 +398,49 @@ function ( $item ) {
+			return ! empty( $recent_product_ids ) ? $recent_product_ids : array();
+		}
+
+		if ( isset( $location['type'] ) && 'cart' === $location['type'] ) {
+			$user_id    = isset( $location['sourceData']['userId'] ) ? absint( $location['sourceData']['userId'] ) : null;
+			$user_email = isset( $location['sourceData']['userEmail'] ) ? sanitize_email( $location['sourceData']['userEmail'] ) : null;
+			if ( $user_id || $user_email ) {
+				return CartCheckoutUtils::get_cart_product_ids_for_user( $user_id, $user_email );
+			}
+		}
```
I think this check is not needed in this case.

lysyjan replied:
I think the method `CartCheckoutUtils::get_cart_product_ids_for_user()` does not find the user and returns an empty array when `user_id` is empty.

costasovo reviewed (approved):
@lysyjan Thank you for addressing the empty state 👍 

As for the rendering issue, I have a basic Woo dev env, the TT5 theme, and I added the block to a Woo transactional email.
See the video.
Note, I approve the PR as this rendering problem may be only on my setup and most probably not related to this PR.

https://github.com/user-attachments/assets/57d789b6-c621-48fa-80fc-2e76e19a68b0

lysyjan commented:
@costasovo I was not able to replicate your problem. I am going to merge it and create follow-up tasks.
